src/NumericalAlgorithms/Spectral/Spectral.cpp

```diff
@@ -321,6 +322,89 @@ const Matrix& linear_filter_matrix(const size_t num_points) noexcept {
       LinearFilterMatrixGenerator<BasisType, QuadratureType>>(num_points);
 }
 
+template <size_t Dim>
```

Comment:
```
namespace {
template <typename F, size_t Dim>
decltype(auto) get_spectral_quantity(F&& f, const Mesh<Dim>& mesh,
                                     const size_t d) noexcept {
  const auto num_points = mesh.extents(d);
  // Switch on runtime values of basis and quadrature to select corresponding
  // template specialization. We do the switch here because with basis
  // functions spanning multiple dimensions there could be more logic involved
  // later.
  switch (mesh.basis(d)) {
    case Basis::Legendre:
      switch (mesh.quadrature(d)) {
        case Quadrature::Gauss:
          return f(std::integral_constant<Basis, Basis::Legendre>{},
                   std::integral_constant<Quadrature, Quadrature::Gauss>{},
                   num_points);
          break;
        case Quadrature::GaussLobatto:
          return f(
              std::integral_constant<Basis, Basis::Legendre>{},
              std::integral_constant<Quadrature, Quadrature::GaussLobatto>{},
              num_points);
          break;
      }
      break;
  }
  ERROR("Unknown spectral");
}
}  // namespace

template <size_t Dim>
const DataVector& collocation_points(const Mesh<Dim>& mesh,
                                     const size_t d) noexcept {
  return get_spectral_quantity(
      [](const auto basis, const auto quadrature,
         const size_t num_points) noexcept->const DataVector& {
        return collocation_points<decltype(basis)::value,
                                  decltype(quadrature)::value>(num_points);
      },
      mesh, d);
}

template <size_t Dim>
const DataVector& quadrature_weights(const Mesh<Dim>& mesh,
                                     const size_t d) noexcept {
  return get_spectral_quantity(
      [](const auto basis, const auto quadrature,
         const size_t num_points) noexcept->const DataVector& {
        return quadrature_weights<decltype(basis)::value,
                                  decltype(quadrature)::value>(num_points);
      },
      mesh, d);
}

template <size_t Dim>
const Matrix& differentiation_matrix(const Mesh<Dim>& mesh,
                                     const size_t d) noexcept {
  return get_spectral_quantity(
      [](const auto basis, const auto quadrature,
         const size_t num_points) noexcept->const Matrix& {
        return differentiation_matrix<decltype(basis)::value,
                                      decltype(quadrature)::value>(num_points);
      },
      mesh, d);
}

template <size_t Dim>
const Matrix& linear_filter_matrix(const Mesh<Dim>& mesh,
                                   const size_t d) noexcept {
  return get_spectral_quantity(
      [](const auto basis, const auto quadrature,
         const size_t num_points) noexcept->const Matrix& {
        return linear_filter_matrix<decltype(basis)::value,
                                    decltype(quadrature)::value>(num_points);
      },
      mesh, d);
}
```

Tested this on your branch